File: challenge5.cpp
//Challenge 5

#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include <sstream>

using namespace std;

// Convert a string to a vector of bytes
vector<uint8_t> string_to_bytes(const string& str) {
    return vector<uint8_t>(str.begin(), str.end());
}

// XOR the plaintext with the repeating key
vector<uint8_t> repeating_key_xor(const vector<uint8_t>& plaintext, const vector<uint8_t>& key) {
    vector<uint8_t> result;
    size_t key_len = key.size();
    
    for (size_t i = 0; i < plaintext.size(); ++i) {
        // XOR the i-th byte of plaintext with the corresponding key byte
        result.push_back(plaintext[i] ^ key[i % key_len]);
    }
    
    return result;
}

// Convert a vector of bytes to a hex string
string bytes_to_hex(const vector<uint8_t>& bytes) {
    stringstream hex_stream;
    for (auto byte : bytes) {
        hex_stream << hex << setw(2) << setfill('0') << (int)byte;
    }
    return hex_stream.str();
}

int main() {
    // Define the plaintext and the key
    string plaintext = "Burning 'em, if you ain't quick and nimble\nI go crazy when I hear a cymbal";
    string key = "ICE";

    // Convert the plaintext and key to byte vectors
    vector<uint8_t> plaintext_bytes = string_to_bytes(plaintext);
    vector<uint8_t> key_bytes = string_to_bytes(key);

    // Encrypt the plaintext using the repeating-key XOR
    vector<uint8_t> encrypted = repeating_key_xor(plaintext_bytes, key_bytes);

    // Convert the encrypted result to a hex string and print it
    string encrypted_hex = bytes_to_hex(encrypted);
    cout << "Encrypted (hex): " << encrypted_hex << endl;

    return 0;
}

File: challenge5.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

class Program
{
    // Convert a hex string to a list of bytes
    static List<byte> HexToBytes(string hex)
    {
        List<byte> bytes = new List<byte>();
        for (int i = 0; i < hex.Length; i += 2)
        {
            string byteValue = hex.Substring(i, 2);
            bytes.Add(Convert.ToByte(byteValue, 16));
        }
        return bytes;
    }

    // Convert a plain text string to a list of bytes (hex format)
    static List<byte> StringToBytes(string text)
    {
        return Encoding.UTF8.GetBytes(text).ToList();
    }

    // XOR the plaintext with the repeating key
    static List<byte> RepeatingKeyXor(List<byte> plaintext, List<byte> key)
    {
        List<byte> result = new List<byte>();
        int keyLen = key.Count;

        for (int i = 0; i < plaintext.Count; i++)
        {
            result.Add((byte)(plaintext[i] ^ key[i % keyLen]));
        }

        return result;
    }

    // Convert a list of bytes to a hex string
    static string BytesToHex(List<byte> bytes)
    {
        StringBuilder hex = new StringBuilder(bytes.Count * 2);
        foreach (byte b in bytes)
        {
            hex.AppendFormat("{0:x2}", b);
        }
        return hex.ToString();
    }

    static void Main()
    {
        // Get the text input from the user
        Console.Write("Enter the text (plain text or hex): ");
        string input = Console.ReadLine();

        // Get the key from the user
        Console.Write("Enter the key: ");
        string key = Console.ReadLine();

        // Check if input is hex or plain text
        List<byte> plaintextBytes;
        if (input.All(c => "0123456789abcdefABCDEF".Contains(c)) && input.Length % 2 == 0)
        {
            // Treat input as hex
            plaintextBytes = HexToBytes(input);
        }
        else
        {
            // Treat input as plain text
            plaintextBytes = StringToBytes(input);
        }

        // Convert key to byte list
        List<byte> keyBytes = Encoding.UTF8.GetBytes(key).ToList();

        // Encrypt the plaintext using the repeating-key XOR
        List<byte> encrypted = RepeatingKeyXor(plaintextBytes, keyBytes);

        // Convert the encrypted result to a hex string and print it
        string encryptedHex = BytesToHex(encrypted);
        Console.WriteLine("Encrypted (hex): " + encryptedHex);
    }
}
